### DISCUSSION TITLE
Design Doc — New Testing Architecture implementation

### DISCUSSION THREAD

**Discussion by kasperski95** [software-mansion/protostar#1370] [Ideas]
## Problem
The New Protostar Testing Architecture has inherited design flaws, including ambiguous naming conventions and questionable code division. To prevent cyclical dependencies, proper code division is crucial. One specific issue is the concept of `Cheaters`, which manage part of the state and provide access to standard StarkNet functionality. Additionally, the term `Cheatcode` has lost its significance. Ambiguous naming also affects the documentation.

## Proposal
![protostar](https://user-images.githubusercontent.com/9629577/212721026-547a4d33-5430-484c-aa1f-340fc660e3fc.png)

Legend:
- blue area — code that will probably remain after switching a test runner
- yellow area — code that will be removed after switching a test runner
- orange area — fuzz-related functionality which is not worth adapting, unless the Rust implementation of Cairo Test Runner further than closer on the timeline

Each subtitle defines an Object Stereotype.
https://khalilstemmler.com/articles/object-oriented/design/object-stereotypes/

Some less important connections are skipped on the diagram to favor clarity.

## Unknowns
Will the Data Transformation logic be reimplemented in Rust?

**Comment by Arcticae** [comment]
This roughly makes sense to me, and conforms to what we discussed, although i have some questions:
- Why the need for InterfacersFactory? Will this factory encapsulate some dependencies and be passed from the DI container down the structure? I am just curious why TestExecutionEnvironment, or TestRunner can't instantiate it directly.
- I don't also understand the purpose of `CheatcodeInterfacer` or `CheatableStarkNetInterfacer` - why the proxy object between `CheatableStarknet` or `Cheater` is needed?

**Reply by kasperski95** [reply]
- InterfacersFactory is similar to CheatcodesFactory. InterfacersFactory will vary at a different rate compared to TestExecutionEnvironment. They can't be instantiated in TestRunner because they will depend on TestExecutionState which is instantiated later, for each test case. That factory (or factories) can instantiate Controllers and provide them to Interfacers as dependencies.
- CheatcodeInterfacer represent a collection of concrete Interfacers classes (~the diagram doesn't reflect that~). Interfacers are responsible for adapting incoming and outgoing data to and from Controllers (CheatableStarknet and Cheater). Controllers are unaware of Interfacers. Theoretically, once the Test Runner in Rust is ready, Views could be replaced and Controllers reused.

**Comment by MaksymilianDemitraszek** [comment]
Good job, I think it represents the idea, issues I see for now are:
- We don't want CheatableStarkNet anymore, we want operate directly on the state
- Interfacer is a weird name (it is a weapon from borderlands, not an proper english word i think 😄 ), can we refrain from using it? I think controller is a solid name don't see a reason why not to use it.
- ProtostarTestingInterfacer -> TestRunner, it is not the same category of enities as the other interfacers
- AssertionControler will depend on CheatableState, we need to check calls in the runtime
- We don't want Cheaters object, we only want CheatcodesController
- Rust Test runner/vm is not in scope of this redesing, please exclude this from the graph

**Reply by kasperski95** [reply]
> We don't want CheatableStarkNet anymore, we want operate directly on the state

`CheatableStarkNet` it's not a `ForkableStarknet`. It's more like current `ContractsCheater` but stateless.

> Interfacer is a weird name (it is a weapon from borderlands, not an proper english word i think 😄 ), can we refrain from using it? I think controller is a solid name don't see a reason why not to use it.

The `Interfacer` is a valid name. It's one of the six Object Stereotypes, which is a concept from Responsibility-Driven Design. A `Controller` is also one of the six Object Stereotypes, but it represents a completely different concept.
(As an alternative to "Interfacer", ChatGPT recommends "Adapter".)

> The last of the stereotypes is the Interfacer. The Interfacer primary acts as a bridge between object neighborhoods (internal and external).
> Internal interfacer: To cross internal object neighborhoods, we need to make the entry-point for our components as simple as possible. The object which sits at the front, relaying messages to objects behind the scenes is the Interfacer.
> External interfacer (gateway): To cross external object neighborhoods (as is the case when we send messages out to remote APIs or to libraries which we don’t own), the more common name for the object that encapsulates this access is called a Gateway, however — stereotypically, it’s an Interfacer.
> UI-related interfacers: Interfacers which listen to events from the UI and pass them to non-UI related parts of our applications are typically just called Event Listeners (in JavaScript, there is an EventListener API) whereas objects that format data for presentation in the UI are aptly named Presenters. Both are stereotypical Interfacers.
https://khalilstemmler.com/articles/object-oriented/design/object-stereotypes/

---

> ProtostarTestingInterfacer -> TestRunner, it is not the same category of enities as the other interfacers

`ProtostarTestingInterfacer` is an abstraction over all `Interfacers`. I should've shown that this is an abstract class on the diagram.

> AssertionControler will depend on CheatableState, we need to check calls in the runtime

~You're right. StarkNet related assertion's state should be in the `CheatableCachedState`.~ I think assertion's state should be kept in `TestExecutionState`.`CheatableCachedState` should store/provide a list of calls.

> We don't want Cheaters object, we only want CheatcodesController

Ok, it can be renamed to `CheatcodesController` to improve naming consistency.

> Rust Test runner/vm is not in scope of this redesing, please exclude this from the graph

This part has lowered opacity for that reason. It gives more concrete argument for the separation between I/O and non I/O code (which is a good thing to do anyway).

**Reply by MaksymilianDemitraszek** [reply]
> CheatableStarkNet it's not a ForkableStarknet. It's more like current ContractsCheater but stateless.

I would prefer to stick to one layer of abstractions, any form of adapters there is unnecessary

**Reply by MaksymilianDemitraszek** [reply]
> The Interfacer is a valid name. It's one of the six Object Stereotypes, which is a concept from Responsibility-Driven Design. 

I see, yea it is valid as language is a dynamic thing, but I would prefer to stick to more common names as probably most people in the team haven't read the book you mentioned 

> ProtostarTestingInterfacer is an abstraction over all Interfacers. I should've shown that this is an abstract class on the diagram.

Makes sense 👍 

> I think assertion's state should be kept in TestExecutionState.CheatableCachedState should store/provide a list of calls.

I considered this as well, this would be cleaner, if you feel the same I am 👍 

>This part has lowered opacity for that reason. It gives more concrete argument for the separation between I/O and non I/O code (which is a good thing to do anyway).

Please remove `Rust Implementation` from the graph, as it is misleading, this is a undecided implementation detail atm

Also please remove a layer of interfacers, and stick only to controllers, we do not need this extra layer of abstraction for our use.

Can you please apply those changes to the graph so those are easily accessible for everyone without reading the whole discussion? 🙏

**Reply by kasperski95** [reply]
> I see, yea it is valid as language is a dynamic thing, but I would prefer to stick to more common names as probably most people in the team haven't read the book you mentioned

I try to use names defined in the industry. Can you suggest an alternative name?

> Also please remove a layer of interfacers, and stick only to controllers, we do not need this extra layer of abstraction for our use.

We need to convert data from and to hints. It's more a [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns). It's basically a  `CairoCheatcode`.

> Please remove Rust Implementation from the graph, as it is misleading, this is a undecided implementation detail atm

Fine, I'll remove it later.

**Reply by MaksymilianDemitraszek** [reply]
Discussed on slack, please rename interfacers to HintLocalView
